### PR TITLE
Add the capability of Downloading Mediation Policies and Role based restricting the Unlimited Tier 

### PR DESCRIPTION
**Pull request by menakaj**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

- Downloading Mediation Policies
Fix wso2/product-apim#2786
- Role-based restricting the Unlimited Subscription Tier. 
Fix wso2/product-apim#2790

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 1]
```diff
@@ -5384,4 +5387,180 @@ protected String getSearchQuery(String searchQuery) throws APIManagementExceptio
...
+                        APIConstants.API_CUSTOM_SEQUENCE_TYPE_FAULT);
+            } else if ("out".equals(type)) {
+                customSeqFileLocation = APIUtil.getSequencePath(identifier,
+                        APIConstants.API_CUSTOM_SEQUENCE_TYPE_OUT);
+            } else {
+                customSeqFileLocation = APIUtil.getSequencePath(identifier,
+                        APIConstants.API_CUSTOM_SEQUENCE_TYPE_IN);
+            }
+
+            if(registry.resourceExists(customSeqFileLocation))    {
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 2]
```diff
@@ -5384,4 +5387,180 @@ protected String getSearchQuery(String searchQuery) throws APIManagementExceptio
...
+                defaultSequenceFileLocation = APIConstants.API_CUSTOM_FAULTSEQUENCE_LOCATION;
+            } else if ("out".equals(type)) {
+                defaultSequenceFileLocation = APIConstants.API_CUSTOM_OUTSEQUENCE_LOCATION;
+            } else {
+                defaultSequenceFileLocation = APIConstants.API_CUSTOM_INSEQUENCE_LOCATION;
+            }
+            if (registry.resourceExists(defaultSequenceFileLocation)) {
+                org.wso2.carbon.registry.api.Collection defaultSeqCollection =
+                        (org.wso2.carbon.registry.api.Collection) registry.get(defaultSequenceFileLocation);
+                if (defaultSeqCollection !=null) {
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 3]
```diff
@@ -1903,6 +1904,26 @@ public boolean isDuplicateContextTemplate(String contextTemplate) throws APIMana
...
+        APIManagerConfiguration apiManagerConfiguration = ServiceReferenceHolder.getInstance()
+                .getAPIManagerConfigurationService().getAPIManagerConfiguration();
+        ThrottleProperties throttleProperties = apiManagerConfiguration.getThrottleProperties();
+
+        if (!throttleProperties.isEnableUnlimitedTier()) {
+            List<Policy> policiesWithoutUnlimitedTier = new ArrayList<Policy>();
+
+            if (policies != null) {
+                for (Policy policy : policies) {
+                    if (!"unlimited".equalsIgnoreCase(policy.getPolicyName())) {
```
Comment (tharikaGitHub) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: Can this be moved to a constant? ("unlimited")

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 4]
```diff
@@ -5384,4 +5387,180 @@ protected String getSearchQuery(String searchQuery) throws APIManagementExceptio
...
+                }
+
+                //Convert the content stream to a string.
+                if (requiredSequence != null) {
+                    sequenceStream = requiredSequence.getContentStream();
+                    StringWriter stringWriter = new StringWriter();
+                    IOUtils.copy(sequenceStream, stringWriter);
+                    sequenceText = stringWriter.toString();
+                } else {
+                    log.error("No sequence for the name " + name + "is found!");
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: you will need space before is found

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 5]
```diff
@@ -5384,4 +5387,180 @@ protected String getSearchQuery(String searchQuery) throws APIManagementExceptio
...
+                }
+            } else {
+                log.error("Invalid arguments.");
+            }
+        }  catch (APIManagementException e) {
+            log.error(e.getMessage());
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: catching thrtowwing the same exception

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 6]
```diff
@@ -5384,4 +5387,180 @@ protected String getSearchQuery(String searchQuery) throws APIManagementExceptio
...
+            throw new APIManagementException(e);
+        } catch (RegistryException e) {
+            log.error(e.getMessage());
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: you might add meaningful message to log

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 7]
```diff
@@ -5384,4 +5387,180 @@ protected String getSearchQuery(String searchQuery) throws APIManagementExceptio
...
+                    }
+                }
+            }
+        }  catch (RegistryException e) {
+            throw new APIManagementException("Error while retrieving registry for tenant " + tenantId, e);
+        } catch (org.wso2.carbon.registry.api.RegistryException e) {
+            throw new APIManagementException("Error while processing the " + defaultSequenceFileLocation +
+                    " in the registry", e);
+        } catch (Exception e) {
+            throw new APIManagementException(e.getMessage(), e);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: add message when you log

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 8]
```diff
@@ -5384,4 +5387,180 @@ protected String getSearchQuery(String searchQuery) throws APIManagementExceptio
...
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+            } else {
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain
+                        (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME, true);
+            }
+            UserRegistry registry = ServiceReferenceHolder.getInstance().getRegistryService()
+                    .getGovernanceSystemRegistry(tenantId);
+
+            String customSeqFileLocation = "";
+            if ("fault".equals(type)) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: better to move these strings to constants

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 9]
```diff
@@ -5384,4 +5387,180 @@ protected String getSearchQuery(String searchQuery) throws APIManagementExceptio
...
+
+    /**
+     * Method to get the user specified mediation sequence.
+     *
+     * @param apiIdentifier : The identifier of the api.
+     * @param type : Mediation type. {in, out, fault}
+     * @param name : The name of the sequence that needed.
+     * @return : The content of the mediation sequence.
+     */
+    public String getSequenceFile(APIIdentifier apiIdentifier, String type, String name) throws APIManagementException {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: sequence file or sequecne string?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 10]
```diff
@@ -5384,4 +5387,180 @@ protected String getSearchQuery(String searchQuery) throws APIManagementExceptio
...
+
+        Resource requiredSequence;
+        InputStream sequenceStream;
+        String sequenceText = "";
+
+        try {
+            if (apiIdentifier != null && type != null && name != null) {
+                requiredSequence = getDefaultSequence(type, name);
+                if (log.isDebugEnabled()) {
+                    log.debug("Get the default sequences.");
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: may be we can add more context to this log saying what is the sequence name?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 5 above)
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Please log with the error

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 4 above)
Comment (menakaj) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Fixed in https://github.com/wso2/carbon-apimgt/pull/5114/commits/726f591adc11d8584c2c1e375a6a5add578d90ed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 9 above)
Comment (menakaj) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Changed to getSequenceFileContent in commit https://github.com/wso2/carbon-apimgt/pull/5114/commits/c6bc99c70b7ef3df0181e7565b77152ab729b066

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 10 above)
Comment (menakaj) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Added more verbose log in commit https://github.com/wso2/carbon-apimgt/pull/5114/commits/c6bc99c70b7ef3df0181e7565b77152ab729b066

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 2 above)
Comment (menakaj) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Fixed in https://github.com/wso2/carbon-apimgt/pull/5114/commits/c6bc99c70b7ef3df0181e7565b77152ab729b066

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 8 above)
Comment (menakaj) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Fixed in https://github.com/wso2/carbon-apimgt/pull/5114/commits/c6bc99c70b7ef3df0181e7565b77152ab729b066

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 1 above)
Comment (menakaj) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Fixed in https://github.com/wso2/carbon-apimgt/pull/5114/commits/c6bc99c70b7ef3df0181e7565b77152ab729b066

**Pull request merged**